### ISSUE TITLE
Implement CMS Update Logic According to Equation 31

### ISSUE THREAD

**Issue by engichang1467** [engichang1467/Open-HOPE#4]
# Verify and Implement Continuum Memory System (CMS) Update Rule (Equation 31)

The Continuum Memory System (CMS) replaces the traditional Feed-Forward Network (FFN). It is defined as a chain of MLP blocks where the update frequency of the $l$-th MLP is governed by the chunk size $C^{(l)}$.

The implementation of the update logic must strictly adhere to **Equation 31** from the paper:

$$\theta^{(f\ell)}_{i+1} = \theta^{(f\ell)}_{i} - \begin{cases} \sum_{t=i-C^{(\ell)}}^{i} \eta^{(l)}_{t} f(\theta^{(f\ell)}_{t}; x_{t}) & \text{if } i \equiv 0 \pmod{C^{(\ell)}}, \\ 0 & \text{otherwise}. \end{cases}$$

This ensures that the parameters $\theta^{(f\ell)}_{t}$ are updated only at the specified frequency intervals, allowing them to compress their context into abstract knowledge. The configuration file (`config/hope_config.yaml`) defines these frequency levels and associated `chunk_size` values, which must be accurately mapped to $C^{(\ell)}$.

**Key Equations:** Equation 30 (forward pass) and Equation 31 (update logic).

**FYI**: Here's Equation 30 below

$$y_t = \text{MLP}^{(f_k)}(\text{MLP}^{(f_{k-1})}(\cdots \text{MLP}^{(f_1)}(x_t)))$$

**Code Location:** `src/memory/cms.py`

**Comment by engichang1467** [comment]
<html><head></head><body><p>The Continuum Memory System (CMS) replaces the traditional Feed-Forward Network (FFN) in the HOPE architecture and is central to the Nested Learning (NL) paradigm. Implementing the CMS update rule (Equation 31) involves defining a conditional, time-aggregated update mechanism based on the concept of <strong>Multi-Time Scale Updates</strong>.</p>
<p>Here is a detailed breakdown of how to implement the CMS Update Rule (Equation 31):</p>
<h3>1. The Core Conditional Update Mechanism</h3>
<p>The CMS is structured as a chain of MLP blocks, $MLP^{(f_1)}(\cdot), \ldots, MLP^{(f_k)}(\cdot)$. The parameters of the $\ell$-th MLP block, denoted $\theta^{(f\ell)}$, are updated only when the time step $i$ is an exact multiple of its assigned chunk size, $C^{(\ell)}$.</p>
<p>The update rule is:</p>
<p>$$\theta^{(f\ell)}<em>{i+1} = \theta^{(f\ell)}</em>{i} - \begin{cases} \sum_{t=i-C^{(\ell)}}^{i} \eta^{(l)}<em>{t} f(\theta^{(f\ell)}</em>{t}; x_{t}) &amp; \text{if } i \equiv 0 \pmod{C^{(\ell)}}, \ 0 &amp; \text{otherwise}. \end{cases}$$</p>
<p>This formula governs the update of parameters $\theta^{(f\ell)}$ from step $i$ to $i+1$.</p>
<h4>Case 1: The Update Occurs (Synchronization Point)</h4>
<p>If the current global step $i$ satisfies the condition $i \equiv 0 \pmod{C^{(\ell)}}$, the parameters are updated.</p>
<ol>
<li><strong>Frequency Check:</strong> This condition ensures that the parameters $\theta^{(f\ell)}$ are updated only at the specified frequency intervals, which are determined by the chunk size $C^{(\ell)}$.</li>
<li><strong>Aggregation:</strong> The update term involves a <strong>summation</strong> over the immediate past $C^{(\ell)}$ steps (from $t=i-C^{(\ell)}$ to $t=i$). This aggregation acts as a mechanism where the parameters $\theta^{(f\ell)}_{t}$ compress their context flow into abstract knowledge.</li>
<li><strong>The Error Component ($f(\cdot)$):</strong> The function $f(\theta^{(f\ell)}<em>{t}; x</em>{t})$ represents the error component of an arbitrary optimizer. In the context of standard optimization like Gradient Descent (GD), $f(\cdot)$ would typically correspond to the gradient, $\nabla L(\theta^{(f\ell)}<em>{t}; x</em>{t})$.</li>
<li><strong>Learning Rate:</strong> $\eta^{(l)}_{t}$ is the learning rate corresponding to the specific parameter $\theta^{(f\ell)}$ at time $t$.</li>
</ol>
<h4>Case 2: No Update Occurs (Between Synchronization Points)</h4>
<p>If the condition $i \equiv 0 \pmod{C^{(\ell)}}$ is <strong>not</strong> met, the update term is zero.</p>
<ol>
<li><strong>Stasis:</strong> This means the parameters $\theta^{(f\ell)}$ remain frozen during that time step: $\theta^{(f\ell)}<em>{i+1} = \theta^{(f\ell)}</em>{i}$.</li>
<li><strong>Implementation Implication:</strong> In a typical training loop, the gradients accumulated for this specific layer during the current step $i$ must be stored or ignored, and the weight update operation must be skipped for this parameter group.</li>
</ol>
<h3>2. Implementation Context and Configuration</h3>
<p>The CMS implementation is located in <code>src/memory/cms.py</code>. The system is built as a chain of MLP blocks, represented internally, for example, as a <code>nn.ModuleList</code> of MLPs.</p>
<h4>Multi-Time Scale Updates</h4>
<p>The design of the CMS emphasizes <strong>Multi-Time Scale Updates</strong>, a fundamental principle of Nested Learning (NL).</p>
<ul>
<li>NL decomposes the architecture into "levels" based on the speed of parameter updates.</li>
<li>In practice, this means the overall training loop (e.g., in <code>src/utils/trainer.py</code>) <strong>cannot use a standard <code>optimizer.step()</code></strong> to update all parameters simultaneously every iteration.</li>
<li>The trainer must track the global step ($i$) and use the <code>chunk_size</code> values ($C^{(\ell)}$) to determine which specific parameter group (level) is eligible for an update at that moment.</li>
</ul>
<h4>Mapping Configuration to $C^{(\ell)}$</h4>
<p>The specific chunk sizes ($C^{(\ell)}$) that govern the update frequency for each MLP level are defined in the configuration file (<code>config/hope_config.yaml</code>).</p>
<p>For example, a configuration might define three levels (low to high frequency):</p>

Name | Frequency (relative to unit time step) | Chunk Size ($C^{(\ell)}$) | Update Behavior
-- | -- | -- | --
level_1 | 1 | 16 | Updates every step (High Freq)
level_2 | 16 | 1000 | Updates every 16 steps
level_3 | 1000 | 0 | Updates rarely (Pre-training knowledge)


<p>The chunk size $C^{(\ell)}$ is the value used in the modulo check ($i \equiv 0 \pmod{C^{(\ell)}}$) and defines the window size for aggregation.</p>
<h3>3. Forward Pass Context (Equation 30)</h3>
<p>Although the query focuses on the update rule (Equation 31), the CMS functions within the HOPE model via its forward pass (Equation 30), which shows that the MLP blocks are chained sequentially:</p>
<p>$$y_t = \text{MLP}^{(f_k)}(\text{MLP}^{(f_{k-1})}(\cdots \text{MLP}^{(f_1)}(x_t)))$$</p>
<p>The output $y_t$ at time $t$ is calculated by passing the input $x_t$ through the sequence of memory layers, where the innermost layer $MLP^{(f_1)}$ has the highest frequency/fastest update, and the outermost layer $MLP^{(f_k)}$ has the lowest frequency/slowest update.</p>
<hr>
<p><strong>Analogy:</strong> The Continuum Memory System (CMS) operating under Equation 31 is like a series of specialized cash registers in a large department store, each handling transactions at a different speed. The "highest frequency" register (Level 1) updates its daily sales log after every single transaction, while the "low frequency" register (Level 3) only aggregates and records its sales totals once every month. Each register only adjusts its long-term parameter (total profits, $\theta$) at its scheduled interval ($i \equiv 0 \pmod{C^{(\ell)}}$), and when it does update, it summarizes the activity ($f(\cdot)$) that occurred since its last reporting period ($C^{(\ell)}$ steps), rather than reacting to every single customer immediately. This differential updating allows the fastest registers to handle immediate context, while the slowest ones gradually condense accumulated context into abstract, long-term knowledge.</p></body></html>
